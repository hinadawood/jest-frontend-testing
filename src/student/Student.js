import React, { Component } from "react";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button/Button";
import { saveStudentInfo } from "./StudentActions";
import { connect } from "react-redux";
import "./student.css";

export class Student extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: props.firstName || "",
      lastName: props.lastName || "",
      age: props.age || ""
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleClickSave = this.handleClickSave.bind(this);
  }

  handleChange(e) {
    this.setState({
      [e.target.name]: e.target.value
    });
  }

  handleClickSave() {
    this.props.saveStudentInfo(
      this.state.firstName,
      this.state.lastName,
      this.state.age
    )
  }

  render() {
    return (
      <div className="page-container">
        <h1>Student Form</h1>
        <form className="my-form">
          <TextField
            name="firstName"
            onChange={e => this.handleChange(e)}
            label="First Name"
            value={this.state.firstName}
            variant="outlined"
            style={{ marginBottom: "10px" }}
          />
          <TextField
            name="lastName"
            onChange={e => this.handleChange(e)}
            label="Last Name"
            value={this.state.lastName}
            variant="outlined"
            style={{ marginBottom: "10px" }}
          />
          <TextField
            name="age"
            onChange={e => this.handleChange(e)}
            label="Age"
            value={this.state.age}
            variant="outlined"
            style={{ marginBottom: "10px" }}
          />
          <Button
            size="large"
            onClick={() => this.handleClickSave()}
            style={{ background: "blue", color: "white"}}
          >
            Save
          </Button>
        </form>
      </div>
    );
  }
}

export const mapStateToProps = state => {
  return {
    student: state.student
  };
};

export const mapDispatchToProps = dispatch => {
  return {
    saveStudentInfo: (firstName, lastName, age) => {
      dispatch(saveStudentInfo(firstName, lastName, age));
    }
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Student);
