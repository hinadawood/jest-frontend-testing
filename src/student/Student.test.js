import React from "react";
import { shallow } from "enzyme";
import { Student } from "./Student";
import renderer from 'react-test-renderer';


let component;

const buildComponent = (props = {
  firstName: 'Jane',
  lastName: 'Doe',
  age: '25',
  saveStudentInfo: jest.fn()
}) => {
  component = shallow(<Student {...props} />);
  return component;
};

describe('test suite for student form', () => {
  beforeEach(() => {
    buildComponent();
  });

  it('should render title properly', () => {
    expect(component).toBeDefined();
    expect(component.text().indexOf('Student Form')).toBeGreaterThan(-1);
  });
});

describe('test change event for student form firstName', () => {
  beforeEach(() => {
    buildComponent();
    component.find('TextField').at(0).simulate('change', {target: {value: 'Mary', name: 'firstName'}})
  });

  it('should update firstName', () => {
    expect(component.state().firstName).toEqual('Mary')
  })
});

describe('test change event for student form lastName', () => {
  beforeEach(() => {
    buildComponent();
    component.find('TextField').at(1).simulate('change', {target: {value: 'Le', name: 'lastName'}})
  });

  it('should update firstName', () => {
    expect(component.state().lastName).toEqual('Le')
  })
});

describe('test change event for student form age', () => {
  beforeEach(() => {
    buildComponent();
    component.find('TextField').at(2).simulate('change', {target: {value: '30', name: 'age'}})
  });

  it('should update firstName', () => {
    expect(component.state().age).toEqual('30')
  })
});

describe('test click event for submit button', () => {
  beforeEach(() => {
    buildComponent();
    component.find('WithStyles(Button)').at(0).simulate('click')
  });

  it('should call saveStudentInfo', () => {
    expect(component.instance().props.saveStudentInfo).toHaveBeenCalled()
  })
});

describe('snapshot testing', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Student/>).toJSON();
    expect(tree).toMatchSnapshot();
  })
})
