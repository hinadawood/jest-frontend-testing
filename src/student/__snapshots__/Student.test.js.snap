// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing renders correctly 1`] = `
<div
  className="page-container"
>
  <h1>
    Student Form
  </h1>
  <form
    className="my-form"
  >
    <div
      className="MuiFormControl-root-1"
      style={
        Object {
          "marginBottom": "10px",
        }
      }
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-outlined-15"
        data-shrink={false}
      >
        First Name
      </label>
      <div
        className="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37"
        onClick={[Function]}
      >
        <fieldset
          aria-hidden={true}
          className="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="MuiPrivateNotchedOutline-legend-54"
            style={
              Object {
                "width": 0.01,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
        <input
          aria-invalid={false}
          className="MuiInputBase-input-46 MuiOutlinedInput-input-31"
          disabled={false}
          name="firstName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-1"
      style={
        Object {
          "marginBottom": "10px",
        }
      }
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-outlined-15"
        data-shrink={false}
      >
        Last Name
      </label>
      <div
        className="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37"
        onClick={[Function]}
      >
        <fieldset
          aria-hidden={true}
          className="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="MuiPrivateNotchedOutline-legend-54"
            style={
              Object {
                "width": 0.01,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
        <input
          aria-invalid={false}
          className="MuiInputBase-input-46 MuiOutlinedInput-input-31"
          disabled={false}
          name="lastName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-1"
      style={
        Object {
          "marginBottom": "10px",
        }
      }
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-outlined-15"
        data-shrink={false}
      >
        Age
      </label>
      <div
        className="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37"
        onClick={[Function]}
      >
        <fieldset
          aria-hidden={true}
          className="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="MuiPrivateNotchedOutline-legend-54"
            style={
              Object {
                "width": 0.01,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
        <input
          aria-invalid={false}
          className="MuiInputBase-input-46 MuiOutlinedInput-input-31"
          disabled={false}
          name="age"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root-81 MuiButton-root-55 MuiButton-text-57 MuiButton-flat-60 MuiButton-sizeLarge-79"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "background": "blue",
          "color": "white",
        }
      }
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-56"
      >
        Save
      </span>
      <span
        className="MuiTouchRipple-root-84"
      />
    </button>
  </form>
</div>
`;

exports[`snapshot testing should render correctly by comparing snapshot 1`] = `
<div
  className="page-container"
>
  <h1>
    Student Form
  </h1>
  <form
    className="my-form"
  >
    <div
      className="MuiFormControl-root-1"
      style={
        Object {
          "marginBottom": "10px",
        }
      }
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-outlined-15"
        data-shrink={false}
      >
        First Name
      </label>
      <div
        className="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37"
        onClick={[Function]}
      >
        <fieldset
          aria-hidden={true}
          className="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="MuiPrivateNotchedOutline-legend-54"
            style={
              Object {
                "width": 0.01,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
        <input
          aria-invalid={false}
          className="MuiInputBase-input-46 MuiOutlinedInput-input-31"
          disabled={false}
          name="firstName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-1"
      style={
        Object {
          "marginBottom": "10px",
        }
      }
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-outlined-15"
        data-shrink={false}
      >
        Last Name
      </label>
      <div
        className="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37"
        onClick={[Function]}
      >
        <fieldset
          aria-hidden={true}
          className="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="MuiPrivateNotchedOutline-legend-54"
            style={
              Object {
                "width": 0.01,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
        <input
          aria-invalid={false}
          className="MuiInputBase-input-46 MuiOutlinedInput-input-31"
          disabled={false}
          name="lastName"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-1"
      style={
        Object {
          "marginBottom": "10px",
        }
      }
    >
      <label
        className="MuiFormLabel-root-16 MuiInputLabel-root-5 MuiInputLabel-formControl-10 MuiInputLabel-animated-13 MuiInputLabel-outlined-15"
        data-shrink={false}
      >
        Age
      </label>
      <div
        className="MuiInputBase-root-36 MuiOutlinedInput-root-23 MuiInputBase-formControl-37"
        onClick={[Function]}
      >
        <fieldset
          aria-hidden={true}
          className="MuiPrivateNotchedOutline-root-53 MuiOutlinedInput-notchedOutline-30"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="MuiPrivateNotchedOutline-legend-54"
            style={
              Object {
                "width": 0.01,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
        <input
          aria-invalid={false}
          className="MuiInputBase-input-46 MuiOutlinedInput-input-31"
          disabled={false}
          name="age"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root-81 MuiButton-root-55 MuiButton-text-57 MuiButton-flat-60 MuiButton-sizeLarge-79"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "background": "blue",
          "color": "white",
        }
      }
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-56"
      >
        Save
      </span>
      <span
        className="MuiTouchRipple-root-84"
      />
    </button>
  </form>
</div>
`;
